### PR TITLE
[SPARK-47810][SQL] Replace equivalent expression to <=> in join condition

### PR DESCRIPTION
**Pull request by zml1206**
### What changes were proposed in this pull request?
Replaces `t1.id = t2.id or (t1.id is null and t2.id is null)` to `t1.id <=> t2.id` in join condition.

### Why are the changes needed?
Improve performance.
`t1.id = t2.id or (t1.id is null and t2.id is null)` and `t1.id <=> t2.id` are equivalent in join condition, but the plan will be different and the performance during execution will be completely different. Id is join key in `t1.id <=> t2.id`, in `t1.id = t2.id or (t1.id is null and t2.id is null)` id cannot be used as join key, this will affect join selection and subsequent matching logic.
For example:
```
sql(
      """
        |CREATE OR REPLACE TEMPORARY VIEW t1 AS SELECT * FROM VALUES
        |(1, 1),
        |(null, 2)
        |AS t1(a, b);
        |""".stripMargin)
    sql(
      """
        |CREATE OR REPLACE TEMPORARY VIEW t2 AS SELECT * FROM VALUES
        |(1, 10),
        |(null, 20)
        |AS t2(a, b);
        |""".stripMargin)
    sql("select t1.a,t2.b from t1 left join t2 on t1.a=t2.a or (t1.a is null and t2.a is null)")
```
before PR is `BroadcastNestedLoopJoin`:
```
== Optimized Logical Plan ==
Project [a#6, b#11]
+- Join LeftOuter, ((a#6 = a#10) OR (isnull(a#6) AND isnull(a#10)))
   :- Project [a#6]
   :  +- LocalRelation [a#6, b#7]
   +- LocalRelation [a#10, b#11]
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- Project [a#6, b#11]
   +- BroadcastNestedLoopJoin BuildRight, LeftOuter, ((a#6 = a#10) OR (isnull(a#6) AND isnull(a#10)))
      :- Project [a#6]
      :  +- LocalTableScan [a#6, b#7]
      +- BroadcastExchange IdentityBroadcastMode, [plan_id=26]
         +- LocalTableScan [a#10, b#11]
```
after PR is BroadcastHashJoin:
```
== Optimized Logical Plan ==
Project [a#6, b#11]
+- Join LeftOuter, (a#6 <=> a#10)
   :- Project [a#6]
   :  +- LocalRelation [a#6, b#7]
   +- LocalRelation [a#10, b#11]

== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- Project [a#6, b#11]
   +- BroadcastHashJoin [coalesce(a#6, 0), isnull(a#6)], [coalesce(a#10, 0), isnull(a#10)], LeftOuter, BuildRight, false
      :- Project [a#6]
      :  +- LocalTableScan [a#6, b#7]
      +- BroadcastExchange HashedRelationBroadcastMode(List(coalesce(input[0, int, true], 0), isnull(input[0, int, true])),false), [plan_id=26]
         +- LocalTableScan [a#10, b#11]
```

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Unit test.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by zml1206** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJoinCondition.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * in join condition for better performance.
+ */
+object OptimizeJoinCondition extends Rule[LogicalPlan] with PredicateHelper {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.transformWithPruning(
+    _.containsPattern(JOIN), ruleId) {
+    case join @ Join(left, right, joinType, condition, hint) if condition.nonEmpty =>
+      val predicates = condition.map(splitConjunctivePredicates).get
+      var replace = false
+      val newPredicates = predicates.map {
+        case Or(EqualTo(l, r), And(IsNull(c1), IsNull(c2)))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJoinCondition.scala:36]: Do we need to call `splitConjunctivePredicates` at the beginning? It seems we can just transform the join condition expression and match this pattern of `Or`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJoinCondition.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+          if (l.semanticEquals(c1) && r.semanticEquals(c2))
+            || (l.semanticEquals(c2) && r.semanticEquals(c1)) =>
+          EqualNullSafe(l, r)
+        case Or(And(IsNull(c1), IsNull(c2)), EqualTo(l, r))
+          if (l.semanticEquals(c1) && r.semanticEquals(c2))
+            || (l.semanticEquals(c2) && r.semanticEquals(c1)) =>
+          EqualNullSafe(l, r)
+      })
+      Join(left, right, joinType, newCondition, hint)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJoinCondition.scala:43]: nit: `j.copy(condition = newCondition)`

**Review by kamcheungting-db — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJoinCondition.scala (hunk 1 above)
Comment (anton5798) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJoinCondition.scala:34]: If the target scenario just has `[l = r OR (isnull(l) and isnull(r))]` somewhere in the expression tree, then I think don't need to `splitConjunctivePredicates`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJoinCondition.scala [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import org.apache.spark.sql.catalyst.plans.logical.{Join, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.{JOIN, OR}
+
+/**
+ * Replaces `t1.id is null and t2.id is null or t1.id = t2.id` to `t1.id <=> t2.id`
+ * in join condition for better performance.
+ */
+object OptimizeJoinCondition extends Rule[LogicalPlan] with PredicateHelper {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.transformWithPruning(
```
Comment (anton5798) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJoinCondition.scala:30]: Even though this is a simple rule, we should add a flag guard here.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala [hunk 4]
```diff
@@ -621,4 +621,14 @@ class DataFrameJoinSuite extends QueryTest
     checkAnswer(joined, Row("x", null, null))
     checkAnswer(joined.filter($"new".isNull), Row("x", null, null))
   }
+
+  test("SPARK-47810: replace equivalent expression to <=> in join condition") {
```
Comment (anton5798) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala:625]: Let's also add a summary of plan change (before/after) and quantified performance benefit in the PR description

Comment (zml1206) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala:625]: Added.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJoinCondition.scala (hunk 3 above)
Comment (zml1206) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJoinCondition.scala:30]: Sorry, I don’t understand what kind of flag it is. Can you give me an example? Thank you.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJoinCondition.scala:30]: We can easily disable a QO rule by setting `spark.sql.optimizer.excludedRules`

**Review by agubichev — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**